google/cloud/storage/client.h

```diff
@@ -2260,7 +2260,7 @@ class Client {
    * @param options a list of optional query parameters and/or request headers.
    *     Valid types for this operation include `UserProject`.
    *
-   * @throw std::runtime_error if there is a permanent failure, or if there were
+   * @returns An error status if there is a permanent failure, or if there were
```

Comment:
This documentation says what the function returns in the event of an error, but it never says what it returns if there is no error. Some options:

* Add another `@return` line for the non-error case (if doxygen allows it). This could be confusing.
* Use `@return` *only* for the non-error case. Not OK status's are understood.
* Omit `@return` completely when the return type is obvious from the function name.

I think the second option seems like it's probably the best. Thoughts?

Reply:
I am voting for the third one.  I added some comments at the top because I am not sure `StatusOr` is as well understood outside google.

Reply:
Option 3 is actually my preference as well. But I think my question wasn't clear. 

This function has a `@return` line that documents the StatusOr's error case only. That seems odd to me. Unless the function intends to document something specific about the failed status (like, maybe which status code is returned), then I think we probably don't want to use `@return` to document the error case.

So in this case, if we think the fact that ListNotifications() returns a vector<Notification> is obvious and doesn't need an `@return` line, then I think we may want to just omit this line completely. However, if we think an `@return` line is needed in this documentation, then perhaps it should focus on the expected value, not the error code. Maybe something like this:

` @return A vector of Notification objects, or an error.`

(or even leave of ", or an error", which is implied by the `StatusOr` type)

Thoughts?

Reply:
I think that makes sense in general.  I agree with you that in general we can omit the `@returns` if it is obvious, I also happen to think that it is obvious in this case.

I also agree that when we return `StatusOr<T>` we can omit the `, on an error.`.